src/Validation/URLValidationRESTController.php

```diff
@@ -0,0 +1,355 @@
+						],
+						'required'    => true,
+						'type'        => 'string',
+					],
+					'id'            => [
+						'description' => __( 'Unique identifier for the object.', 'amp' ),
+						'required'    => true,
+						'type'        => 'integer',
+					],
+					'preview_id'    => [
+						'description' => __( 'Unique identifier for the preview.', 'amp' ),
+						'required'    => false,
+						'type'        => 'integer',
+					],
+					'preview_nonce' => [
+						'description' => __( 'Preview nonce string.', 'amp' ),
+						'required'    => false,
+						'type'        => 'string',
+					],
+				],
+				[
+					'args'                => $this->get_endpoint_args_for_item_schema( WP_REST_Server::READABLE ),
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'validate_post_url' ],
+					'permission_callback' => [ $this, 'get_item_permissions_check' ],
+				],
+				'schema' => [ $this, 'get_public_item_schema' ],
+			]
+		);
+
+		// @todo Additional endpoint to validate a URL (from a URL rather than a post ID).
+	}
+
+	/**
+	 * Checks whether the current user can view AMP validation results.
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has permission; WP_Error object otherwise.
+	 */
+	public function get_item_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( ! $this->dev_tools_user_access->is_user_enabled() ) {
+			return new WP_Error(
+				'amp_rest_no_dev_tools',
+				__( 'Sorry, you do not have access to dev tools for the AMP plugin for WordPress.', 'amp' ),
+				[ 'status' => rest_authorization_required_code() ]
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Returns validation information about a URL, validating the URL along the way.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function validate_post_url( $request ) {
+		$post          = get_post( $request['id'] );
+		$preview_id    = $request['preview_id'];
+		$preview_nonce = $request['preview_nonce'];
+		$url           = amp_get_permalink( $post );
+
+		if ( ! empty( $preview_id ) && ! empty( $preview_nonce ) ) {
+			$url = add_query_arg(
+				[
+					'preview'       => 1,
+					'preview_id'    => $preview_id,
+					'preview_nonce' => $preview_nonce,
+				],
+				$url
+			);
+		}
+
+		$this->url_validation_provider->get_url_validation( $url, get_post_type( $post ), true );
+		$validation_status_post = AMP_Validated_URL_Post_Type::get_invalid_url_post( $url );
+
+		$data = [
+			'results'     => [],
+			'review_link' => get_edit_post_link( $validation_status_post->ID, 'raw' ),
+		];
+
+		foreach ( AMP_Validated_URL_Post_Type::get_invalid_url_validation_errors( $validation_status_post ) as $result ) {
+			$data['results'][] = [
+				'error'       => $result['data'],
```

Comment:
To ensure that `$result['data']['sources']` will be an array, should we do the same `array_values()` here at runtime? This would ensure the previously-saved data (bbb0e49 and 7617125) would be updated.

Then again, does the WordPress REST API return `{}` when serializing an empty `array()` as opposed to `[]`?

Reply:
I'm not seeing this as being the case. I tried adding this to the theme's `footer.php` without being inside of any hook:

```
<script>/*No source*/</script>
```

When I look at the validated URL data as returned by the REST API, I see:

```json
{"error":{"type":"js_error","text":"\/*No source*\/","sources":[]}
```

So I'm not seeing empty PHP arrays coming through as empty objects in PHP.

<!--Also, if I try reproducing the issue of duplicate errors by undoing the `array_unique()` then I get-->

Reply:
Done in a91b14a.